### ISSUE TITLE
[pkg/ottl] Ensure arrays in attribute values are homogeneous

### ISSUE THREAD

**Issue by evan-bradley** [open-telemetry/opentelemetry-collector-contrib#16355]
### Component(s)

pkg/ottl

### Is your feature request related to a problem? Please describe.

Arrays in attribute values [are required to be homogeneous](https://opentelemetry.io/docs/reference/specification/common/#attribute) by the specification. While OTLP does support heterogeneous arrays for attribute values, backends are expecting spec-compliant data, so we should ensure the OTTL conforms to the specification where possible.

Logs support [heterogeneous arrays in the body field](https://opentelemetry.io/docs/reference/specification/logs/data-model/#field-body) to support existing log models, so we should still allow heterogeneous arrays in this case. The specification advises log bodies to be strings, so this should only be used when working with specific formats.

### Describe the solution you'd like

Update attribute setters in OTTL contexts to transform any array type that isn't one of the supported primitive types (`string`, `bool`, `int64`, `float64`, or `[]byte`) into an array of one of these types. Any attempts to set an attribute value to a heterogeneous type should return an error, which should ideally be logged for the user to debug.

### Describe alternatives you've considered

We could not validate this, but that would open users to setting invalid attribute values.

**Comment by fatsheep9146** [comment]
May I try this issue? @evan-bradley

**Comment by evan-bradley** [comment]
That would be great. It's yours.

**Comment by fatsheep9146** [comment]
I think this work should be after #16338 is merged, right? @evan-bradley

**Comment by evan-bradley** [comment]
The implementation doesn't have a hard dependency on that PR since the validation should probably be performed in the attributes setters which aren't touched there. However, being able to use lists in Getters will make live testing easier.

**Comment by fatsheep9146** [comment]
After reading through several prs. I have some questions want to make sure before starting coding.
- In order to `Update attribute setters in OTTL contexts`, I think the new code should be in `SetMapValue` function for `pkg/ottl/contexts/internal/ottlcommon/map.go` since the span, metricdatapoint, log attributes setter all call `SetMapValue`.
- `transform any array type that isn't one of the supported primitive types (string, bool, int64, float64, or []byte) into an array of one of these types. Any attempts to set an attribute value to a heterogeneous type should return an error`. Does this means that if we encounter the case that user input a attribute which type is array of heterogeneous type, we should convert them to an array of homogeneous type. for example, user input an array ["aa", 100, "bb"], we should convert it into `["aa", "100", "bb"]`?

@evan-bradley

**Comment by evan-bradley** [comment]
> In order to Update attribute setters in OTTL contexts, I think the new code should be in SetMapValue function for pkg/ottl/contexts/internal/ottlcommon/map.go since the span, metricdatapoint, log attributes setter all call SetMapValue.

Unfortunately I think we will need to put this in the attribute setters, as `SetMapValue` will need to support heterogeneous arrays if it is called from the log body setter. I think a helper function that transforms an `[]any` type array into a primitive-typed array and is used inside the attributes setters before calling `SetMapValue` is likely going to be the cleanest solution possible.

> Does this means that if we encounter the case that user input a attribute which type is array of heterogeneous type, we should convert them to an array of homogeneous type. for example, user input an array ["aa", 100, "bb"], we should convert it into ["aa", "100", "bb"]?

Returning an error instead of converting the type would likely be the best move here, I worry that implicitly converting the elements would be a source of confusion. If users want to perform type conversion on arrays we should offer a factory function that does this so that it's explicit.

**Comment by fatsheep9146** [comment]
Thanks for explanation.

> Returning an error instead of converting the type would likely be the best move here, I worry that implicitly converting the elements would be a source of confusion. If users want to perform type conversion on arrays we should offer a factory function that does this so that it's explicit.

I agree with your point, we should throw out the error, not converting implicitly.

> Unfortunately I think we will need to put this in the attribute setters, as `SetMapValue` will need to support heterogeneous arrays if it is called from the log body setter. I think a helper function that transforms an `[]any` type array into a primitive-typed array and is used inside the attributes setters before calling `SetMapValue` is likely going to be the cleanest solution possible.

So I think I should modify the `accessAttributesKey` function in `pkg/ottl/contexts/internal/ottlcommon/span.go`, like following, is this right?

```go
func accessAttributesKey[K SpanContext](mapKey *string) ottl.StandardGetSetter[K] {
	return ottl.StandardGetSetter[K]{
                 ...
		Setter: func(ctx context.Context, tCtx K, val interface{}) error {

                         // add new validation logic here, if failed return error 

			SetMapValue(tCtx.GetSpan().Attributes(), *mapKey, val)
			return nil
		},
	}
}

```

@evan-bradley

**Comment by evan-bradley** [comment]
>So I think I should modify the accessAttributesKey function in pkg/ottl/contexts/internal/ottlcommon/span.go, like following, is this right?

Yeah, I think that approach is likely the best one available.

**Comment by bogdandrutu** [comment]
I personally believe that we should not enforce this. It seems unnecessary overhead that does not bring that much value to the user. Also the requirement is mostly for OpenTelemetry API/SDKs, but when converting from different formats that may not be necessary a requirement, since the OTLP protocol does NOT require the same thing. See https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/protocol and https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L44

**Comment by evan-bradley** [comment]
> Also the requirement is mostly for OpenTelemetry API/SDKs, but when converting from different formats that may not be necessary a requirement, since the OTLP protocol does NOT require the same thing. 

My understanding was that the Collector should avoid producing data that the APIs or SDKs would not produce, even if the values are representable in OTLP. I see the value in supporting any OTLP data regardless of whether it is valid per the specification, so I'm good with not enforcing this.

I'll leave this issue open for a little in case anyone else has any other thoughts.

**Comment by fatsheep9146** [comment]
@TylerHelmuth any thoughts about this?

**Comment by TylerHelmuth** [comment]
I am ok not enforcing this.  @evan-bradley unless you feel differently I'm good with closing this issue and the PR as "wont fix".

**Comment by evan-bradley** [comment]
Closing as "not planned" since we have consensus on this. Even though it's implicit by the scope of the OpenTelemetry specification, we should probably make it clear somewhere that we're only bound by the OTLP data model (since that is what pdata is based on) and may export OTLP data that wouldn't be ordinarily possible to produce with an OpenTelemetry SDK, but that's a separate issue.
